### ISSUE TITLE
Training multiple xgboost models in parallel uses the same cpus for all the models

### ISSUE THREAD

**Issue by dcarnino** [dmlc/xgboost#2814]
## Problem summary
When training `N` xgboost models (using `nthread` >= 1 for each model) in a multiprocessing environment with `N` processes, each model is trained with the same `nthread` cpus instead of taking advantage of the `N` * `nthread` cpus made available.
For example, when running the code below on a 256 cpus machine, all 10 xgboost models are trained on the same 8 cpus instead of training them with 8 * 10 = 80 distinct cpus. Training 10 models with a single thread for each of them results in having all models trained on the same cpu simultaneously, no matter how many cpus are available on the machine. This was observed with `htop`.
Being able to train multiple xgboost models concurrently on distinct cpus seems like a very important feature to me. It is useful for ensembling, stacking, hyperparameter search, cross-validation and many other applications. One workaround is to sequentially train each model on all cpus, but this results in a massive waste of computation time on machines with >2 cpus as xgboost training time does not scale linearly with the number of cpus used for training.

## Minimal working example
```python
import numpy as np
import multiprocessing as mp
from xgboost import XGBClassifier

def xgb_func(nthread):
    X = np.random.randn(10000,100)
    y = (np.random.random((10000,)) > .5)
    model = XGBClassifier(objective="binary:logistic", nthread=nthread)
    model.fit(X, y)

if __name__ == '__main__':
    N = 10
    nthread = 8
    processes = []

    for ix_trial in range(N):
        process = mp.Process(target=xgb_func, args=(nthread,))
        process.start()
        processes.append(process)

    for process in processes:
        process.join()
```

## Things I have tried
I have tried the following without success. The behaviour is still the same.
1. Setting the `nthread` parameter of `XGBClassifier` to 1. All xgboosts are then trained on the same core.
2. Using `xgboost.train` instead of the sklearn API.
3. Setting multiprocessing context to fork/forkserver/spawn.
4. Trying `sklearn.model_selection.cross_val_score` with `n_jobs`  > 1 and `XGBClassifier` as tested model.

## Environment info
OS: `4.9.0-0.bpo.3-amd64 #1 SMP Debian 4.9.30-2+deb9u2~bpo8+1 (2017-06-27) x86_64 GNU/Linux`
Compiler: `gcc 4.9.2`
Package used: `python`
The python version and distribution: `python 3.6.0` from `anaconda3`
Xgboost version used: `0.6a2`
Installed with: `pip install xgboost`
Also tried installing from source without OpenMP support. Same behaviour.

**Comment by khotilov** [comment]
Do you mean each model is using only one thread when running in parallel (100% in top per process), disregarding nthread? There are some issues regarding gcc's OpenMP implementation and multiprocessing in python. Check #2163 for some background and suggestions.

**Comment by dcarnino** [comment]
Each model is trained with the correct number of threads, but all the threads from the different models use the same cpus. So if 256 models are trained concurrently with `nthread` = 1 for each model, the first cpu will have 100% and the 255 other cpus will have 0% in htop.
I read #2163 before. So I tried different contexts for multiprocessing (fork, forkserver and spawn). And also tried without support for OpenMP (option `make no_omp=1` when building). Neither of the suggestions solved the problem.
Have you tried the minimal working example, see if it is possible to replicate the problem?

**Comment by khotilov** [comment]
I've tried your example in python 3.5 on an idle 16 cores / 32 virtual cores server, with the change of `nthread = 8` and adding `n_estimators=500` to run a bit longer. I saw 10 processes running at 300% of %CPU in top, with the total CPU utilization of 93.8%us = 30/32 as expected. That's actually encouraging, because in the past there would be crashes or hangs when using more than 1 openmp thread in multiprocessing. When using a single thread though it was always working as expected and utilizing Nprocess of CPU resources. 

So I am not sure why multiprocessing of single thread jobs in your case does not run in parallel. Is it really an xgboost-specific issue, i.e., does the multiprocessing work as expected with anything else than xgboost? What is that 256 cores system?

**Comment by dcarnino** [comment]
Thanks for your feedback. It is good to know the problem happens only on my machine.
The 256 cores system is a linux server (`4.9.0-0.bpo.3-amd64 #1 SMP Debian 4.9.30-2+deb9u2~bpo8+1 (2017-06-27) x86_64 GNU/Linux`). The multiprocessing works well with other tasks. For example training `N = 10` scikit-learn random forests and `n_jobs = 20` threads for each random forest uses 200 cores at 100% (*i.e.* 10 processes running at ~2000% of %CPU in `top`).
